### PR TITLE
Refine file locks for inter process communication

### PR DESCRIPTION
**Pull request by gossion**
- [x] Please check this box and fill the data as below once you have run the unit tests.
      Code coverage before your change: `865 examples, 0 failures. 3595 / 3658 LOC (98.28%) covered.`
      Code coverage with your change: `839 examples, 0 failures. 3445 / 3501 LOC (98.4%) covered.`

  ```
  pushd src/bosh_azure_cpi
    ./bin/test-unit
  popd
  ```

  NOTE: Please see how to setup dev environment and run unit tests in docs/development.md.

### Changelog

* Refine file locks for inter process communication, and fix the issue that the empty availability set may not be deleted due to lock issue

### CODE REVIEW

File: src/bosh_azure_cpi/lib/cloud/azure/cloud.rb [hunk 1]
```diff
@@ -695,16 +695,7 @@ def init_azure
...
-        ignore_exception(Errno::EEXIST) { Dir.mkdir(CPI_LOCK_DIR) }
-      elsif needs_deleting_locks?
-        @logger.info('init_cpi_lock_dir: Cleaning up the locks')
-        Dir.glob("#{CPI_LOCK_DIR}/#{CPI_LOCK_PREFIX}*") do |file_name|
-          @logger.debug("init_cpi_lock_dir: Deleting the lock `#{file_name}'")
-          ignore_exception(Errno::ENOENT) { File.delete(file_name) }
-        end
-        ignore_exception(Errno::ENOENT) { remove_deleting_mark }
-      end
+      ignore_exception(Errno::EEXIST) { Dir.mkdir(CPI_LOCK_DIR) } unless Dir.exist?(CPI_LOCK_DIR)
```
Comment (andyliuliming) [inline src/bosh_azure_cpi/lib/cloud/azure/cloud.rb]: minor: better to use the mkdir_p
require 'fileutils'
ignore_exception(Errno::EEXIST) {FileUtils.mkdir_p(path) unless Dir.exists?(path)}

File: src/bosh_azure_cpi/lib/cloud/azure/helpers.rb [hunk 2]
```diff
@@ -544,232 +537,26 @@ def get_os_disk_size(root_disk_size, stemcell_info, use_root_disk)
...
-      rescue Errno::ENOENT => e
-        raise LockNotFoundError
-      end
-
-      def update
-        raise LockNotOwnedError unless @is_locked
+    def flock(lock_name, mode)
+      file_path = "#{CPI_LOCK_DIR}/#{lock_name}"
+      file = File.open(file_path, File::RDWR | File::CREAT, 0o644)
+      success = file.flock(mode)
```
Comment (andyliuliming) [inline src/bosh_azure_cpi/lib/cloud/azure/helpers.rb]: I think flock does not works on some file systems, say the nfs,
if customer mount  :(
it's better to check the file system first.

and actually I think the best solution for the inter-process lock would be the azure storage blob lease.

Comment (gossion) [reply src/bosh_azure_cpi/lib/cloud/azure/helpers.rb]: 1. There is no any document in https://ruby-doc.org/core-2.4.2/File.html#method-i-flock saying that flock is not supported by NFS. And I just had a simple test in my local ubuntu VM, it works well when the file is in a NFS folder.
2. Azure storage blob lease is not a good solution for this case, we need some `exclusive` and `share` locks, but blob lease is always a `exclusive` lock. Another important thing is that we will use the lock in creating storage account, but in some case we can use the blob lease as the storage account does not exist at that time.
3. Won't consider named pipe this time, as flock is good enough.

File: src/bosh_azure_cpi/lib/cloud/azure/cloud.rb (hunk 1 above)
Comment (gossion) [reply src/bosh_azure_cpi/lib/cloud/azure/cloud.rb]: thanks. but `Dir.exists?` is not needed anymore if using `mkdir_p`.

File: src/bosh_azure_cpi/lib/cloud/azure/helpers.rb (hunk 2 above)
Comment (andyliuliming) [reply src/bosh_azure_cpi/lib/cloud/azure/helpers.rb]: https://github.com/rubygems/rubygems/pull/1327

seems the NFS support is added.

**Pull request merged**